### PR TITLE
`lib.lists.{findFirstIndex,commonPrefix}`: init

### PR DESCRIPTION
**Pull request by infinisil**
###### Description of changes

Adds two new functions:
- `lib.lists.findFirstIndex :: (a -> bool) -> Any -> [a] -> (Int | Any)`: Find the first index in the list matching the specified predicate or return a default if no such element exists.
  ```nix
  lib.lists.findFirstIndex (x: x > 3) null [ 0 6 4 ]
  -> 1
  ```
- `lib.lists.commonPrefix :: [ a ] -> [ a ] -> [ a ]`: The common prefix of two lists:
  ```nix
  lib.lists.commonPrefix [ 1 2 3 4 ] [ 1 2 4 8 ]
  -> [ 1 2 ]
  ```

This is distantly part of the [path library effort](https://github.com/NixOS/nixpkgs/issues/210426), in order to be able to work with lists of path components. Relates to https://github.com/NixOS/nixpkgs/pull/243511

This work is sponsored by [Antithesis](https://antithesis.com/) :sparkles:

###### Things done
- [x] Docs
- [x] Tests

### CODE REVIEW

File: lib/lists.nix [hunk 1]
```diff
@@ -637,6 +637,76 @@ rec {
...
+
+  Type: commonPrefix :: [a] -> [a] -> [a]
+
+  Example:
+    commonPrefix [ 1 2 3 4 5 6 ] [ 1 2 4 8 ]
+    => [ 1 2 ]
+    commonPrefix [ 1 2 3 ] [ 1 2 3 4 5 ]
+    => [ 1 2 3 ]
+    commonPrefix [ 1 2 3 ] [ 4 5 6 ]
+    => [ 0 ]
```
Comment (fricklerhandwerk) [inline lib/lists.nix:703]:
```suggestion
    => [ ]
```

File: lib/lists.nix [hunk 2]
```diff
@@ -637,6 +637,76 @@ rec {
...
+
+  Example:
+    commonPrefixLength [ 1 2 3 4 5 6 ] [ 1 2 4 8 ]
+    => 2
+    commonPrefixLength [ 1 2 3 ] [ 1 2 3 4 5 ]
+    => 3
+    commonPrefixLength [ 1 2 3 ] [ 4 5 6 ]
+    => 0
+  */
+  commonPrefixLength =
```
Comment (tjni) [inline lib/lists.nix:652]: Is recursion possible here to simplify? In pseudocode:
```
commonPrefixLength list1 list2 =
    if list1 is empty or list2 is empty or list1[0] != list2[0] then
        0
    else
        1 + commonPrefixLength (tail list1) (tail list2)
```

I read that `tail` is inefficient but a current index can be passed along to emulate it.

Comment (infinisil) [reply lib/lists.nix:652]: This would cause a stack overflow if the list is too long, there's a test for that too. See also https://github.com/NixOS/nixpkgs/pull/235267#discussion_r1212986842 where this was also discussed

Comment (tjni) [reply lib/lists.nix:652]: Sounds good.

File: lib/lists.nix (hunk 1 above)
Comment (fricklerhandwerk) [reply lib/lists.nix:703]: > Return 0 when there is no better value to be found.

File: lib/lists.nix [hunk 3]
```diff
@@ -637,6 +637,76 @@ rec {
...
+    let
+      shorterFirst =
+        shorter: longer:
+        let
+          # We fold over the shorter list, no need to check elements of the
+          # longer list that couldn't be matched by the shorter one
+
+          # index invariant:
+          # - if index < 0 then (- index - 1) is the first index that differs between the two lists
+          # - if index >= 0 then all elements below that index match and s == elemAt shorter index
```
Comment (fricklerhandwerk) [inline lib/lists.nix:662]:
```suggestion
          # - if index >= 0 then all elements below that index match and `s == elemAt shorter index` holds
```

That was hard to parse otherwise

File: lib/lists.nix (hunk 2 above)
Comment (fricklerhandwerk) [reply lib/lists.nix:652]: The comment should say why it is not done this way here. Here's a good reason to be clever, but without explanation it naturally raises the question "why not just..."

File: lib/lists.nix [hunk 4]
```diff
@@ -637,6 +637,76 @@ rec {
...
+        if result < 0 then
+          # Elements at index (- index - 1) differ, which is also the length of the common prefix
+          - result - 1
+        else
+          # We reached the end of the shorter list, indicating that all elements matched the longer one
+          result;
+    in
+    list1: list2:
+    if length list1 < length list2 then
+      shorterFirst list1 list2
```
Comment (fricklerhandwerk) [inline lib/lists.nix:689]:
```suggestion
      commonPrefixLength' list1 list2
```

The other name doesn't really say what it's about.

**Review by fricklerhandwerk — COMMENTED** [review]
I like the minimality of all this, a lot.

**Comment by tjni** [comment]
Just brainstorming here: what do you think about extracting out `findFirst` into `findFirstIndex` and reusing that here? Pseudocode being:
```
commonPrefixLength list1 list2 =
    let pairs = zipLists list 1 list2
    let firstDiffIndex = findFirstIndex ({ fst, snd }: fst != snd) pairs
    if firstDiffIndex was found then
        firstDiffIndex
    else
        length pairs
```
Not sure about the performance of this or if `firstFindIndex` is the best generalization of the `foldl'` idea you have.

**Comment by tjni** [comment]
Thinking about this a little more, the following can also be a route if `commonPrefixLength` is otherwise not too useful by itself (which I am unsure about but think it might be only useful as a helper function for `commonPrefix`):

We use your algorithm and create the more general:
```
findFirstWithIndex = predicate: list:
    return first { index, value } matching predicate
```
and then use that to define:
```
takeWhile = predicate: list:
    let sentinel = findFirstWithIndex (elem: !(predicate elem)) list in
    if sentinel != null then
        take sentinel.index list
    else
        list
```
and finally:
```
commonPrefix = list1: list2:
    takeWhile (predicate comparing list1[i] with list2[i])
```
where forming that last predicate depends on some design decisions (like does `predicate` accept the index, do we use `zipLists`, do we first get the shorter list and take over it explicitly).

**Comment by roberth** [comment]
> `findFirstWithIndex`

This seems like a useful function to factor out.
To simplify it a bit, you could return just the index. That way you don't have to allocate the `{ index, value }` attrset just to ignore `value`. You could call it `findFirstIndex` then.

**Comment by infinisil** [comment]
@tjni Nice idea! This turned out to work out really well since I was able to reuse a lot of code. I removed `commonPrefixLength` again, it's not really necessary to expose that, it's really just a helper function as you suggested. Instead `findFirstIndex` is now available :)

File: lib/lists.nix [hunk 5]
```diff
@@ -180,18 +180,18 @@ rec {
...
 
      Example:
-       findFirst (x: x > 3) 7 [ 1 6 4 ]
-       => 6
-       findFirst (x: x > 9) 7 [ 1 6 4 ]
-       => 7
+       findFirstIndex (x: x > 3) null [ 1 6 4 ]
+       => 1
+       findFirstIndex (x: x > 9) null [ 1 6 4 ]
+       => null
```
Comment (roberth) [inline lib/lists.nix:192]:
```suggestion
       findFirstIndex (x: x > 3) null [ 0 6 4 ]
       => 1
       findFirstIndex (x: x > 9) null [ 0 6 4 ]
       => null
```

Make 1 unambiguous.

**Review by fricklerhandwerk — APPROVED** [review]
Very elegant, such compose, many wow.

**Review by roberth — APPROVED** [review]
nod

**Review by tjni — APPROVED** [review]
Looks awesome!
